referenceframe/linear_inputs.go

```diff
@@ -0,0 +1,188 @@
+package referenceframe
+
+import (
+	"fmt"
+	"iter"
+)
+
+// LinearInputsSchema describes the order of frame names and their degrees of freedom in a set of
+// Inputs. This is important for communicating consistently with IK/nlopt.
+type LinearInputsSchema struct {
+	metas []linearInputMeta
+}
+
+// FloatsToInputs applies the given schema to a new set of linearized floats. This returns an error
+// if the wrong number of floats are provided.
+func (lis *LinearInputsSchema) FloatsToInputs(inps []float64) (*LinearInputs, error) {
+	totDoF := 0
+	for idx := range lis.metas {
+		totDoF += lis.metas[idx].dof
+	}
+	if totDoF != len(inps) {
+		return nil, fmt.Errorf("wrong number of inputs. Expected: %v Received: %v", totDoF, len(inps))
+	}
+
+	return &LinearInputs{
+		schema: lis,
+		inputs: inps,
+	}, nil
+}
+
+// FrameNamesInOrder returns the frame names in schema order.
+func (lis *LinearInputsSchema) FrameNamesInOrder() []string {
+	ret := make([]string, len(lis.metas))
+	for idx, meta := range lis.metas {
+		ret[idx] = meta.frameName
+	}
+
+	return ret
+}
+
+type linearInputMeta struct {
+	frameName string
+
+	// Having both Offset and DoF is merely a convenience to keep information local. Only having one
+	// would suffice. But then we'd need to compare adjacent `linearInputMeta` in the schema array
+	// of `metas`.
+	offset int
+	dof    int
```

Comment:
The "ugly" part is that if I have a "set limits" method -- this `linearInputMeta` has two forms. One where we just know the number of degrees of freedom. And one where we know the actual min/max limits of each DoF

Reply:
@erh  comment #2